trace-dispatcher/doc/trace-dispatcher.md

```diff
@@ -0,0 +1,798 @@
+
+As mentioned above, __dispatcher__ is the point of interpretation of the program traces -- a structured set of __Tracer__ objects, that defines and implements the __language and policy__ of __trace interpretation__.
+
+__Trace interpretation__ is specified in terms of:
+
+* __trace synthesis__, which means production of __synthetic traces__ -- in cases where we decide it is cheaper (or at all possible) to perform trace aggregation inside the program,
+* __trace naming__, which is assignment of hierarchically-structured names to all traces -- which serve identification, documentation and configuration purposes,
+* __trace filtering__: which, in turn relies on notions of __severity__, __privacy__ and __frequency__ of messages,
+* __trace presentation__ : relying on __detail level__ and on the `LogFormatting` transformation of the traces into JSON, human readable and metric forms -- the last step before traces meet their __trace-outs__,
+* __trace documentation__, as a mode of the __dispatcher__ operation.
+
+The __trace interpretation__ process requires that for each traced type the __dispatcher__ is provided with:
+
+* instances of the `LogFormatting` typeclass, and
+* __trace prototypes__ and __trace documentation__.
+
+__Trace interpretation__ would have been unusably static, if it wasn't allowed to be configured without recompilation -- and therefore the __effective tracing policy__ used by the __dispatcher__ can be partially defined by the externally-supplied __trace configuration__.
+
+The __effective tracing policy__ defines for each trace a __trace context__, which is what effectively informs interpretation performed by the __dispatcher__ for that particular trace.
+
+The __trace context__, in turn, consists of the __logging context__ encoded in the __dispatcher__, and the __configuration context__ coming from the __trace configuration__.
+
+As a final note, the __dispatcher__ is not provided by the `trace-dispatcher` library as a ready-made, turn-key component -- instead, we are provided with __trace combinators__, the building blocks that allow its construction -- and therefore, expression of the desirable __trace interpretation policies__.
+
+# Interface overview
+
+## The trace / tracer duality
+
+__Traces__ begin with a definition of their values, or __messages__.  As an example:
+
+```haskell
+data TraceAddBlockEvent blk =
+    IgnoreBlockOlderThanK (RealPoint blk)
+  | IgnoreBlockAlreadyInVolatileDB (RealPoint blk)
+  ...
+```
+
+__Traces__ cannot be entered into the tracing system, unless they are accompanied by a matching __tracer__ -- a monadic callback, that expresses the action of tracing of values of that particular type:
+
+```haskell
+trAddBlock :: Trace IO (TraceAddBlockEvent blk)
```

Comment:
Typo? Is this meant to say `Tracer` i.e. missing an "r"?

Reply:
I see we're adding a new `Trace` type and a new `Tracer` type and keeping around `Tracer` from `contra-tracer`. Is this a temporary naming scheme until we remove `contra-tracer`?

Reply:
We build on the arrow based contra-tracer library. Since the iohk-monitoring framework depends on the non-arrow based contra-tracer framework, we integrated contra-tracer-0.1.. into the source tree.
This should become a dependency later.

Reply:
I'm still slightly confused here (though this is likely because I've only read through documentation and not the code). Why is this type named `Trace` instead of `Tracer`? Is this just to avoid a name clash with `Tracer` from `contra-tracer`?

Reply:
Let's continue this thread here: https://github.com/input-output-hk/cardano-node/pull/3073#discussion_r704604233

Reply:
It's because we augment the `Tracer` type to carry context, and call it `Trace`, which is a name choice that was carried over from the previous system.

This `Trace` is not the same thing as the conceptual "trace", which is an abstract entity.

It's simplest to think of `Trace` as "just a minor variation of `Tracer`".